### PR TITLE
AFS time series storage

### PR DESCRIPTION
**Pull request by geofjamg**

### CODE REVIEW

File: afs/afs-mapdb-storage/src/main/java/com/powsybl/afs/mapdb/storage/TimeSeriesIndexSerializer.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+import java.io.IOException;
+import java.io.Serializable;
+
+/**
+ * @author Geoffroy Jamgotchian <geoffroy.jamgotchian at rte-france.com>
+ */
+public class TimeSeriesIndexSerializer implements Serializer<TimeSeriesIndex>, Serializable {
+
+    public static final TimeSeriesIndexSerializer INSTANCE = new TimeSeriesIndexSerializer();
```
Comment (mathbagu) [inline afs/afs-mapdb-storage/src/main/java/com/powsybl/afs/mapdb/storage/TimeSeriesIndexSerializer.java]: Add a private constructor

File: afs/afs-mapdb-storage/src/main/java/com/powsybl/afs/mapdb/storage/StringArrayChunkSerializer.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+import java.io.IOException;
+import java.io.Serializable;
+
+/**
+ * @author Geoffroy Jamgotchian <geoffroy.jamgotchian at rte-france.com>
+ */
+public class StringArrayChunkSerializer implements Serializer<StringArrayChunk>, Serializable {
+
+    public static final StringArrayChunkSerializer INSTANCE = new StringArrayChunkSerializer();
```
Comment (mathbagu) [inline afs/afs-mapdb-storage/src/main/java/com/powsybl/afs/mapdb/storage/StringArrayChunkSerializer.java]: Add a private constructor

File: afs/afs-mapdb-storage/src/main/java/com/powsybl/afs/mapdb/storage/DoubleArrayChunkSerializer.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+import java.io.IOException;
+import java.io.Serializable;
+
+/**
+ * @author Geoffroy Jamgotchian <geoffroy.jamgotchian at rte-france.com>
+ */
+public class DoubleArrayChunkSerializer implements Serializer<DoubleArrayChunk>, Serializable {
+
+    public static final DoubleArrayChunkSerializer INSTANCE = new DoubleArrayChunkSerializer();
```
Comment (mathbagu) [inline afs/afs-mapdb-storage/src/main/java/com/powsybl/afs/mapdb/storage/DoubleArrayChunkSerializer.java]: Add a private constructor?

File: afs/afs-mapdb-storage/src/main/java/com/powsybl/afs/mapdb/storage/MapDbAppFileSystemStorage.java [hunk 4]
```diff
@@ -67,26 +77,50 @@ public int hashCode() {
...
+                return nodeId.equals(other.nodeId) && name.equals(other.name);
             }
             return false;
         }
     }
 
-    private static final class UnorderedNodeIdPair implements Serializable {
+    public static class NamedLinkSerializer implements Serializer<NamedLink>, Serializable {
+
+        public static final NamedLinkSerializer INSTANCE = new NamedLinkSerializer();
```
Comment (mathbagu) [inline afs/afs-mapdb-storage/src/main/java/com/powsybl/afs/mapdb/storage/MapDbAppFileSystemStorage.java]: Add a private constructor?

File: afs/afs-mapdb-storage/src/main/java/com/powsybl/afs/mapdb/storage/MapDbAppFileSystemStorage.java [hunk 5]
```diff
@@ -103,17 +137,161 @@ public boolean equals(Object obj) {
         }
     }
 
-    private final ConcurrentMap<String, NodeId> rootNodeMap;
+    public static class UnorderedNodeIdPairSerializer implements Serializer<UnorderedNodeIdPair>, Serializable {
+
+        public static final UnorderedNodeIdPairSerializer INSTANCE = new UnorderedNodeIdPairSerializer();
```
Comment (mathbagu) [inline afs/afs-mapdb-storage/src/main/java/com/powsybl/afs/mapdb/storage/MapDbAppFileSystemStorage.java]: Add a private constructor

File: afs/afs-mapdb-storage/src/main/java/com/powsybl/afs/mapdb/storage/MapDbAppFileSystemStorage.java [hunk 6]
```diff
@@ -505,89 +719,242 @@ public void close() throws IOException {
...
+    }
+
+    @Override
+    public void removeAllTimeSeries(NodeId nodeId) {
+        UuidNodeId uuidNodeId = checkNodeId(nodeId);
+        Set<String> names = timeSeriesNamesMap.get(uuidNodeId);
+        if (names != null) {
+            for (String name : names) {
+                timeSeriesMetadataMap.remove(new NamedLink(uuidNodeId, name));
+            }
```
Comment (mathbagu) [inline afs/afs-mapdb-storage/src/main/java/com/powsybl/afs/mapdb/storage/MapDbAppFileSystemStorage.java]: names.forEach(remove)?

File: afs/afs-mapdb-storage/src/main/java/com/powsybl/afs/mapdb/storage/TimeSeriesMetadataSerializer.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * @author Geoffroy Jamgotchian <geoffroy.jamgotchian at rte-france.com>
+ */
+public class TimeSeriesMetadataSerializer implements Serializer<TimeSeriesMetadata>, Serializable {
+
+    public static final TimeSeriesMetadataSerializer INSTANCE = new TimeSeriesMetadataSerializer();
```
Comment (mathbagu) [inline afs/afs-mapdb-storage/src/main/java/com/powsybl/afs/mapdb/storage/TimeSeriesMetadataSerializer.java]: Add a private constructor?

File: afs/afs-mapdb-storage/src/main/java/com/powsybl/afs/mapdb/storage/UuidNodeIdList.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    public int size() {
+        return nodeIds.size();
+    }
+
+    public UuidNodeId get(int i) {
+        return nodeIds.get(i);
+    }
+
+    public List<NodeId> getNodeIds() {
+        return nodeIds.stream().collect(Collectors.toList());
```
Comment (mathbagu) [inline afs/afs-mapdb-storage/src/main/java/com/powsybl/afs/mapdb/storage/UuidNodeIdList.java]: Probably efficency issue: to clone a list, use clone or a copy-contructor?

File: afs/afs-mapdb-storage/src/main/java/com/powsybl/afs/mapdb/storage/UuidNodeIdListSerializer.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * @author Geoffroy Jamgotchian <geoffroy.jamgotchian at rte-france.com>
+ */
+public class UuidNodeIdListSerializer implements Serializer<UuidNodeIdList>, Serializable {
+
+    public static final UuidNodeIdListSerializer INSTANCE = new UuidNodeIdListSerializer();
```
Comment (mathbagu) [inline afs/afs-mapdb-storage/src/main/java/com/powsybl/afs/mapdb/storage/UuidNodeIdListSerializer.java]: Add a private constructor?

File: afs/afs-mapdb-storage/src/main/java/com/powsybl/afs/mapdb/storage/UuidNodeIdListSerializer.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+        out.writeInt(nodeIdList.size());
+        for (int i = 0; i < nodeIdList.size(); i++) {
+            UuidNodeIdSerializer.INSTANCE.serialize(out, nodeIdList.get(i));
+        }
+    }
+
+    @Override
+    public UuidNodeIdList deserialize(DataInput2 input, int available) throws IOException {
+        int size = input.readInt();
+        List<UuidNodeId> nodeIds = new ArrayList<>();
```
Comment (mathbagu) [inline afs/afs-mapdb-storage/src/main/java/com/powsybl/afs/mapdb/storage/UuidNodeIdListSerializer.java]: initialize the list with a capacity

File: afs/afs-storage-api/pom.xml [hunk 11]
```diff
@@ -44,13 +44,23 @@
...
         <dependency>
             <groupId>junit</groupId>
             <artifactId>junit</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava-testlib</artifactId>
+            <scope>test</scope>
+        </dependency>
```
Comment (mathbagu) [inline afs/afs-storage-api/pom.xml]: This dependency should be before junit (alphabetic order)

File: math/src/main/java/com/powsybl/math/timeseries/AbstractCompressedArrayChunk.java [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+
+    protected final int offset;
+
+    protected final int uncompressedLength;
+
+    protected final int[] stepLengths;
+
+    public AbstractCompressedArrayChunk(int offset, int uncompressedLength, int[] stepLengths) {
+        this.offset = offset;
+        this.uncompressedLength = uncompressedLength;
```
Comment (mathbagu) [inline math/src/main/java/com/powsybl/math/timeseries/AbstractCompressedArrayChunk.java]: Check for positive values (>=0)

File: math/src/main/java/com/powsybl/math/timeseries/AbstractPoint.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+ * @author Geoffroy Jamgotchian <geoffroy.jamgotchian at rte-france.com>
+ */
+public abstract class AbstractPoint {
+
+    protected final int index;
+
+    protected final long time;
+
+    protected AbstractPoint(int index, long time) {
+        this.index = index;
```
Comment (mathbagu) [inline math/src/main/java/com/powsybl/math/timeseries/AbstractPoint.java]: Check for positive values

File: math/src/main/java/com/powsybl/math/timeseries/AbstractTimeSeries.java [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+
+    public Stream<P> stream() {
+        return fillGap().stream().flatMap(chunk -> chunk.stream(metadata.getIndex()));
+    }
+
+    public Iterator<P> iterator() {
+        return Iterators.concat(fillGap().stream().map(c -> c.iterator(metadata.getIndex())).collect(Collectors.toList()).iterator());
+    }
+
+    public void writeJson(JsonGenerator generator) {
```
Comment (mathbagu) [inline math/src/main/java/com/powsybl/math/timeseries/AbstractTimeSeries.java]: Add a null-check ?

File: math/src/main/java/com/powsybl/math/timeseries/AbstractCompressedArrayChunk.java [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+
+    protected abstract int getUncompressedEstimatedSize();
+
+    public double getCompressionFactor() {
+        return ((double) getEstimatedSize()) / getUncompressedEstimatedSize();
+    }
+
+    protected abstract void writeStepValuesJson(JsonGenerator generator) throws IOException;
+
+    public void writeJson(JsonGenerator generator) {
```
Comment (mathbagu) [inline math/src/main/java/com/powsybl/math/timeseries/AbstractCompressedArrayChunk.java]: Add a null-check?

File: math/src/main/java/com/powsybl/math/timeseries/AbstractUncompressedArrayChunk.java [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+ * @author Geoffroy Jamgotchian <geoffroy.jamgotchian at rte-france.com>
+ */
+public abstract class AbstractUncompressedArrayChunk {
+
+    private static final double COMPRESSION_FACTOR = 1d;
+
+    protected final int offset;
+
+    public AbstractUncompressedArrayChunk(int offset) {
+        this.offset = offset;
```
Comment (mathbagu) [inline math/src/main/java/com/powsybl/math/timeseries/AbstractUncompressedArrayChunk.java]: Check for positive values?

File: math/src/main/java/com/powsybl/math/timeseries/AbstractUncompressedArrayChunk.java [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+        return false;
+    }
+
+    public double getCompressionFactor() {
+        return COMPRESSION_FACTOR;
+    }
+
+    protected abstract void writeValuesJson(JsonGenerator generator) throws IOException;
+
+    public void writeJson(JsonGenerator generator) {
```
Comment (mathbagu) [inline math/src/main/java/com/powsybl/math/timeseries/AbstractUncompressedArrayChunk.java]: Assert generator is not null

File: math/src/main/java/com/powsybl/math/timeseries/CompressedStringArrayChunk.java [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+        return uncompressedEstimatedSize;
+    }
+
+    @Override
+    public TimeSeriesDataType getDataType() {
+        return TimeSeriesDataType.STRING;
+    }
+
+    @Override
+    public void fillArray(String[] array) {
```
Comment (mathbagu) [inline math/src/main/java/com/powsybl/math/timeseries/CompressedStringArrayChunk.java]: Ensure that array is big enough?

File: math/src/main/java/com/powsybl/math/timeseries/DoublePoint.java [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+            DoublePoint other = (DoublePoint) obj;
+            return index == other.index && time == other.time
+                    && ((Double.isNaN(value) && Double.isNaN(other.value)) || value == other.value);
+        }
+        return false;
+    }
+
+    @Override
+    public String toString() {
+        return "DoublePoint(" + index + ", " + Instant.ofEpochMilli(time) + ", " + value + ")";
```
Comment (mathbagu) [inline math/src/main/java/com/powsybl/math/timeseries/DoublePoint.java]: Add parameter's name?

File: math/src/main/java/com/powsybl/math/timeseries/DoubleTimeSeries.java [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+        }
+
+        class DoublePointExt {
+
+            private final DoublePoint point;
+
+            private final int timeSeriesNum;
+
+            DoublePointExt(DoublePoint point, int timeSeriesNum) {
+                this.point = point;
```
Comment (mathbagu) [inline math/src/main/java/com/powsybl/math/timeseries/DoubleTimeSeries.java]: Assert for non-null point.

File: math/src/main/java/com/powsybl/math/timeseries/RegularTimeSeriesIndex.java [hunk 21]
```diff
@@ -0,0 +1,167 @@
...
+    private final long spacing; // in ms
+
+    private final int firstVersion;
+
+    private final int versionCount;
+
+    public RegularTimeSeriesIndex(long startTime, long endTime, long spacing, int firstVersion, int versionCount) {
+        this.startTime = startTime;
+        this.endTime = endTime;
+        this.spacing = spacing;
```
Comment (mathbagu) [inline math/src/main/java/com/powsybl/math/timeseries/RegularTimeSeriesIndex.java]: Check for positive values

File: math/src/main/java/com/powsybl/math/timeseries/RegularTimeSeriesIndex.java [hunk 22]
```diff
@@ -0,0 +1,167 @@
...
+                    endTime == otherIndex.endTime &&
+                    spacing == otherIndex.spacing &&
+                    firstVersion == otherIndex.firstVersion &&
+                    versionCount == otherIndex.versionCount;
+        }
+        return false;
+    }
+
+    @Override
+    public void writeJson(JsonGenerator generator) {
```
Comment (mathbagu) [inline math/src/main/java/com/powsybl/math/timeseries/RegularTimeSeriesIndex.java]: Assert generator is not null

File: math/src/main/java/com/powsybl/math/timeseries/StringPoint.java [hunk 23]
```diff
@@ -0,0 +1,46 @@
...
+        if (obj instanceof StringPoint) {
+            StringPoint other = (StringPoint) obj;
+            return index == other.index && time == other.time && Objects.equals(value, other.value);
+        }
+        return false;
+    }
+
+    @Override
+    public String toString() {
+        return "StringPoint(" + index + ", " + Instant.ofEpochMilli(time) + ", " + value + ")";
```
Comment (mathbagu) [inline math/src/main/java/com/powsybl/math/timeseries/StringPoint.java]: Add parameter's name

File: math/src/main/java/com/powsybl/math/timeseries/TimeSeries.java [hunk 24]
```diff
@@ -0,0 +1,209 @@
...
+
+    List<C> getChunks();
+
+    Stream<P> stream();
+
+    Iterator<P> iterator();
+
+    void writeJson(JsonGenerator generator);
+
+    static void writeJson(JsonGenerator generator, List<TimeSeries> timeSeriesList) {
```
Comment (mathbagu) [inline math/src/main/java/com/powsybl/math/timeseries/TimeSeries.java]: Assert generator and timeSeriesList are not null

File: math/src/main/java/com/powsybl/math/timeseries/TimeSeriesMetadata.java [hunk 25]
```diff
@@ -0,0 +1,169 @@
...
+
+    public Map<String, String> getTags() {
+        return tags;
+    }
+
+    public TimeSeriesIndex getIndex() {
+        return index;
+    }
+
+    public void writeJson(JsonGenerator generator) {
```
Comment (mathbagu) [inline math/src/main/java/com/powsybl/math/timeseries/TimeSeriesMetadata.java]: Assert generator is not null

File: pom.xml [hunk 26]
```diff
@@ -413,6 +414,11 @@
                 <version>${mapdb.version}</version>
             </dependency>
             <dependency>
+                <groupId>org.threeten</groupId>
+                <artifactId>threeten-extra</artifactId>
+                <version>${threeten.version}</version>
+            </dependency>
```
Comment (mathbagu) [inline pom.xml]: This dependency is not well sorted: in fact the two previous dependencies too :/

File: afs/afs-mapdb-storage/src/main/java/com/powsybl/afs/mapdb/storage/UuidNodeIdList.java (hunk 8 above)
Comment (geofjamg) [reply afs/afs-mapdb-storage/src/main/java/com/powsybl/afs/mapdb/storage/UuidNodeIdList.java]: stream & map to downcast UuidNodeId -> NodeId

File: math/src/main/java/com/powsybl/math/timeseries/AbstractCompressedArrayChunk.java (hunk 12 above)
Comment (geofjamg) [reply math/src/main/java/com/powsybl/math/timeseries/AbstractCompressedArrayChunk.java]: See check method just after called in Compressed*ArrayChunk

File: math/src/main/java/com/powsybl/math/timeseries/DoubleTimeSeries.java (hunk 20 above)
Comment (geofjamg) [reply math/src/main/java/com/powsybl/math/timeseries/DoubleTimeSeries.java]: Useless, internal class

File: math/src/main/java/com/powsybl/math/timeseries/AbstractCompressedArrayChunk.java (hunk 15 above)
Comment (geofjamg) [reply math/src/main/java/com/powsybl/math/timeseries/AbstractCompressedArrayChunk.java]: Done

File: math/src/main/java/com/powsybl/math/timeseries/AbstractTimeSeries.java (hunk 14 above)
Comment (geofjamg) [reply math/src/main/java/com/powsybl/math/timeseries/AbstractTimeSeries.java]: Done

File: math/src/main/java/com/powsybl/math/timeseries/CompressedStringArrayChunk.java (hunk 18 above)
Comment (geofjamg) [reply math/src/main/java/com/powsybl/math/timeseries/CompressedStringArrayChunk.java]: Done

File: math/src/main/java/com/powsybl/math/timeseries/RegularTimeSeriesIndex.java (hunk 22 above)
Comment (geofjamg) [reply math/src/main/java/com/powsybl/math/timeseries/RegularTimeSeriesIndex.java]: Done

File: math/src/main/java/com/powsybl/math/timeseries/TimeSeries.java (hunk 24 above)
Comment (geofjamg) [reply math/src/main/java/com/powsybl/math/timeseries/TimeSeries.java]: Done

File: math/src/main/java/com/powsybl/math/timeseries/TimeSeriesMetadata.java (hunk 25 above)
Comment (geofjamg) [reply math/src/main/java/com/powsybl/math/timeseries/TimeSeriesMetadata.java]: Done

**Pull request merged**